### DISCUSSION TITLE
lint squiggles do not appear in vscode with remark extension installed

### DISCUSSION THREAD

**Discussion by LokiMidgard** [remarkjs/remark#1363] [Q&A] [answered]
I installed remark extension to vscode and the formatting works, but warnings are not displayed (not in the problem window, nor in the editor)

I added following package.json
```json
{
  "name": "sand-artefakte",
  "version": "1.0.0",
  "description": "",
  "scripts": {
    "format": "remark . --output"
  },
  "author": "",
  "license": "ISC",
  "remarkConfig": {
    "plugins": [
      "remark-preset-lint-consistent",
      "remark-preset-lint-recommended",
      "remark-preset-lint-markdown-style-guide",
      "remark-frontmatter",
      "remark-lint"
    ], 
    "settings": {
      "bullet": "-"
    }
  },
  "dependencies": {
    "remark": "^15.0.1",
    "remark-cli": "^12.0.1",
    "remark-frontmatter": "^5.0.0",
    "remark-lint": "^10.0.0",
    "remark-lint-maximum-line-length": "^4.0.1",
    "remark-preset-lint-consistent": "^6.0.0",
    "remark-preset-lint-markdown-style-guide": "^6.0.0",
    "remark-preset-lint-recommended": "^7.0.0",
    "remark-toc": "^9.0.0"
  }
}
```

executing `npm run format` outputs 

```
…
12:1          warning Unexpected unordered list marker `*`, expected `-`
          unordered-list-marker-style   remark-lint
13:1          warning Unexpected unordered list marker `*`, expected `-`
          unordered-list-marker-style   remark-lint
…
```

however the editor shows nothing:
![image](https://github.com/user-attachments/assets/8c70f361-e019-4b65-a300-7fa2ed724a84)

**Comment by remcohaszing** [comment]
I am seeing diagnostics with the code that you provided.

![VSCode showing remark diagnostics](https://github.com/user-attachments/assets/b20736d7-0d13-4821-a0b7-e70b47e3038e)

You dependencies are up-to-date, so that’s good. What version of VSCode and the remark extension are you using? And what operating system?

**Reply by LokiMidgard** [reply]
Thanks for the reply

The VSCode version is:
```
Version: 1.93.0 (user setup)
Commit: 4849ca9bdf9666755eb463db297b69e5385090e3
Date: 2024-09-04T13:02:38.431Z
Electron: 30.4.0
ElectronBuildId: 10073054
Chromium: 124.0.6367.243
Node.js: 20.15.1
V8: 12.4.254.20-electron.0
OS: Windows_NT x64 10.0.22631
```

I just set up everything for remark so the version is `v3.0.0` and should be the current I guess.

I also tried to disable a bunch of extensions to see if there was a conflict, but no success so far…

**Reply by remcohaszing** [reply]
Those versions all look good. Windows support can be tricky sometimes, but it should work. Do you use WSL by any chance?

**Reply by LokiMidgard** [reply]
I noticed that when I create another file in the project, it actually worked.

I tried to find out what Part of the original file was responsable for the error. And the result was this:
[outline.md](https://github.com/user-attachments/files/16932658/outline.md)

- Removing any line from this will trigger the display of errors
- If you add something again, it will still show errors, only after a reload of VSCode errors disapear again
- I started with a file with more text and removing the quote block triggerd it earlier then removing other blocks
- Also removing long lines seem to trigger it earlier

It could be a memory problem, so the file may not reproduce this behavior on another system…

**Reply by LokiMidgard** [reply]
> Those versions all look good. Windows support can be tricky sometimes, but it should work. Do you use WSL by any chance?

No, I used it in the past, but have disabled it now (not sure if it was ever enabled on this Laptop)

**Reply by remcohaszing** [reply, marked as answer]
I managed to reproduce it by running the following script in the `vscode-remark` repo.

```js
import {PassThrough} from 'node:stream'
import {engine} from 'unified-engine'
import {VFile} from 'vfile'
import {remark} from 'remark'

const lorem =
  'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum. No dicam aperiam vis. An dicant apeirian qui, at vide indoctum pro. Offendit eleifend moderatius ex vix, quem odio mazim et qui, purto expetendis cotidieque quo cu, veri persius vituperata ei nec. Pri posse graeco definitiones cu, id eam populo quaestio adipiscing, usu quod malorum te. Et clita interesset quo. Scripta periculis ei eam, te pro movet reformidans. Offendit eleifend moderatius ex vix, quem odio mazim et qui, purto expetendis cotidieque quo cu, veri persius vituperata ei nec. Scripta periculis ei eam, te pro movet reformidans. Offendit eleifend moderatius ex vix, quem odio mazim et qui, purto expetendis cotidieque quo cu, veri persius vituperata ei nec. Offendit eleifend moderatius ex vix, quem odio mazim et qui, purto expetendis cotidieque quo cu, veri persius vituperata ei nec. Eos verear percipit ex, eos ne eligendi inimicus.\n'

const value = lorem.repeat(10)

await new Promise((resolve) => {
  engine(
    {
      files: [
        new VFile({
          value,
          path: new URL('example.md', import.meta.url)
        })
      ],
      packageField: 'remarkConfig',
      pluginPrefix: 'remark',
      processor: remark,
      streamError: new PassThrough(),
      streamOut: new PassThrough()
    },
    (errors, code, context) => {
      console.dir(errors)
      console.dir(code)
      console.dir(context)
      resolve(undefined)
    }
  )
})
```

When this script is run using Node.js 20.15.1 (the version bundled with VSCode) or 20.17.0 (latest in the 20.x release line), this yields exit code 13. Exit code 13 means a promise didn’t get resolved. If this is run using Node.js 22, it runs the console logging and exists succesfully.

I did some debugging, but I didn’t manage to find out what causes it.

**Reply by LokiMidgard** [reply]
Thank you for your help. I just updated Node.js :)

**Reply by remcohaszing** [reply]
Note that I’m explicitly talking about the Node.js version that comes with VSCode. You don’t control that. The problem will eventually go away with some VSCode update that updates to Node.js 22, but I have no clue when that would be.

Since this is an issue in `unified-engine`, caused by a regression in Node.js 20, perhaps @wooorm has an idea how to fix it.

**Reply by wooorm** [reply]
What is the problem?

**Reply by remcohaszing** [reply]
The callback given to `unified-engine` in the script above is never called when using Node.js 20. It works fine in Node.js 22.

**Reply by wooorm** [reply]
You are saying that example does not print anything right? I see 3 things printed on `v20.15.1`.

**Reply by wooorm** [reply]
(edited my above comment)

**Reply by remcohaszing** [reply]
Right, I see 3 things printed when using Node.js 22. When using Node.js 20, I see nothing printed and the exit code is 13. I ran it from the `vscode-remark` repo originally, but I get the same result if I run it from a freshly cloned `unified` repo.

Increasing the number passed to `repeat` to increase the input size, may cause the issue to appear.

**Reply by wooorm** [reply]
`node example.js; echo $?` also returns `0` on `v20.15.1`

**Reply by wooorm** [reply]
package.json:

```json
{
  "name": "example",
  "version": "1.0.0",
  "type": "module",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "description": "",
  "dependencies": {
    "remark": "^15.0.1",
    "unified-engine": "^11.2.1",
    "vfile": "^6.0.3"
  }
}
```

`example.js`:

```js
import {PassThrough} from 'node:stream'
import {engine} from 'unified-engine'
import {VFile} from 'vfile'
import {remark} from 'remark'

const lorem =
  'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum. No dicam aperiam vis. An dicant apeirian qui, at vide indoctum pro. Offendit eleifend moderatius ex vix, quem odio mazim et qui, purto expetendis cotidieque quo cu, veri persius vituperata ei nec. Pri posse graeco definitiones cu, id eam populo quaestio adipiscing, usu quod malorum te. Et clita interesset quo. Scripta periculis ei eam, te pro movet reformidans. Offendit eleifend moderatius ex vix, quem odio mazim et qui, purto expetendis cotidieque quo cu, veri persius vituperata ei nec. Scripta periculis ei eam, te pro movet reformidans. Offendit eleifend moderatius ex vix, quem odio mazim et qui, purto expetendis cotidieque quo cu, veri persius vituperata ei nec. Offendit eleifend moderatius ex vix, quem odio mazim et qui, purto expetendis cotidieque quo cu, veri persius vituperata ei nec. Eos verear percipit ex, eos ne eligendi inimicus.\n'

const value = lorem.repeat(10000)

await new Promise((resolve) => {
  engine(
    {
      files: [
        new VFile({
          value,
          path: new URL('example.md', import.meta.url)
        })
      ],
      packageField: 'remarkConfig',
      pluginPrefix: 'remark',
      processor: remark,
      streamError: new PassThrough(),
      streamOut: new PassThrough()
    },
    (errors, code, context) => {
      console.dir(errors)
      console.dir(code)
      resolve(undefined)
    }
  )
})
```

(note the large repeated number, and no context printed).

```sh
$ node example.js; echo $?
undefined
0
0
```

**Reply by remcohaszing** [reply]
Well 10000 should definitely do it. Maybe it’s an OS specific issue. I’ll try it on Windows.

@LokiMidgard what operating system are you using?

**Reply by remcohaszing** [reply]
I’m seeing the problem too in Windows 11 using Node.js 20.7.0.

**Reply by LokiMidgard** [reply]
> Well 10000 should definitely do it. Maybe it’s an OS specific issue. I’ll try it on Windows.
> 
> @LokiMidgard what operating system are you using?

I also had the problem on Windows 11

**Reply by LokiMidgard** [reply]
```
Version: 1.93.1 (user setup)
Commit: 38c31bc77e0dd6ae88a4e9cc93428cc27a56ba40
Date: 2024-09-11T17:20:05.685Z
Electron: 30.4.0
ElectronBuildId: 10073054
Chromium: 124.0.6367.243
Node.js: 20.15.1
V8: 12.4.254.20-electron.0
OS: Windows_NT x64 10.0.22631
```
